src/Init/Data/BitVec/Bitblast.lean

```diff
@@ -430,6 +430,394 @@ theorem shiftLeft_eq_shiftLeftRec (x : BitVec w₁) (y : BitVec w₂) :
+    apply Nat.div_eq_of_lt n.isLt
+}
+
+/--
+A lawful DivModState with a fully consumed dividend (`wn = 0`) witneses that the
+quotient has been correctly computed.
+-/
+theorem DivModState.udiv_eq_of_lawful_zero {qr : DivModState w}
+    (h : DivModState.Lawful w w 0 n d qr) :
+    n.udiv d = qr.q := by
+  apply udiv_eq_of_mul_add_toNat h.hdPos h.hrLtDivisor
+  have hdiv := h.hdiv
+  omega
+
+/--
+A lawful DivModState with a fully consumed dividend (`wn = 0`) witneses that the
+remainder has been correctly computed.
+-/
+theorem DivModState.umod_eq_of_lawful_zero {qr : DivModState w}
+    (h : DivModState.Lawful w w 0 n d qr) :
+    n.umod d = qr.r := by
+  apply umod_eq_of_mul_add_toNat h.hrLtDivisor
+  have hdiv := h.hdiv
+  simp only [shiftRight_zero] at hdiv
+  exact hdiv.symm
+
+/-! ### LawfulShiftSubtract -/
+
+/--
+A `LawfulShiftSubtract` is a `Lawful` DivModState that is also a legal input to the shift subtractor.
+So in particular, we must have at least one dividend bit left over `(0 < wn)`
+to perform a round of shift subtraction.
+
+The input to the shift subtractor is a legal input to `divrem`, and we also need to have an
+input bit to perform shift subtraction on, and thus we need `0 < wn`.
+-/
+structure DivModState.LawfulShiftSubtract (w wr wn : Nat) (n d : BitVec w) (qr : DivModState w)
+  extends DivModState.Lawful w wr wn n d qr : Type where
+  /-- Only perform a round of shift-subtract if we have dividend bits. -/
+  hwn_lt : 0 < wn
+
+/--
+In the shift subtract input, the dividend is at least one bit long (`wn > 0`), so
+the remainder has bits to be computed (`wr < w`).
+-/
+def DivModState.wr_lt_w {qr : DivModState w} (h : qr.LawfulShiftSubtract wr wn n d) : wr < w := by
+  have hwrn := h.hwrn
+  have hwn_lt := h.hwn_lt
+  omega
+
+/-- If we have extra bits to spare in `n`,
+then the div rem input can be converted into a shift subtract input
+to run a round of the shift subtracter. -/
+def DivModState.Lawful.toLawfulShiftSubtract {qr : DivModState w}
+    (h : qr.Lawful w wr (wn + 1) n d) : qr.LawfulShiftSubtract wr (wn + 1) n d where
+  hwrn := by have := h.hwrn; omega
+  hdPos := h.hdPos
+  hrLtDivisor := h.hrLtDivisor
+  hrWidth := h.hrWidth
+  hqWidth := h.hqWidth
+  hdiv := h.hdiv
+  hwn_lt := by omega
+
+/-! ### shiftConcat -/
+
+/--
+If `n : Bitvec w` has only the low `k < w` bits set,
+then `(n <<< 1 | b)` does not overflow.
+-/
+private theorem mul_two_add_lt_two_pow_of_lt_two_pow_of_lt_two {n b k w : Nat}
+    (hn : n < 2 ^ k) (hb : b < 2) (hk : k < w) :
+    n * 2 + b < 2 ^ w := by
+  have : 2^(k + 1) ≤ 2 ^w := Nat.pow_le_pow_of_le_right (by decide) (by assumption)
+  omega
+
+@[simp, bv_toNat]
+theorem toNat_shiftConcat {x : BitVec w} {b : Bool} : (x.shiftConcat b).toNat =
+    (x.toNat <<< 1 + b.toNat) % 2 ^ w  := by
+  simp only [shiftConcat]
+  rw [← add_eq_or_of_and_eq_zero] -- Due to `add_eq_or_of_and_eq_zero`, this must live in `Bitblast`.
+  · simp
+  · ext i
+    simp
+    omega
```

Comment:
I would prefer these `toNat_shiftConcat` theorems to be in `BitVec/Lemmas.lean`. I guess the proof would be a bit more involved, but maybe that's worth knowing that including `Lemmas.lean` gives you a complete toNat interface.

Reply:
It's not possible for it live in `Lemmas.lean` at the moment, because it depends on being able to rewrite bitwise `or` into addition. These theorems are proven based on the `adc` definition that lives in `Bitblast.lean`.